phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJsonDataType.java

```diff
@@ -0,0 +1,258 @@
+
+		if (!actualType.isCoercibleTo(this)) {
+			throwConstraintViolationException(actualType, this);
+		}
+		if (length == 0) {
+			return null;
+		}
+		return getPhoenixJson(bytes, offset, length);
+
+	}
+
+	@Override
+	public Object toObject(Object object,
+			@SuppressWarnings("rawtypes") PDataType actualType) {
+		if (object == null) {
+			return null;
+		}
+		if (equalsAny(actualType, PJsonDataType.INSTANCE)) {
+			return object;
+		}
+		if (equalsAny(actualType, PVarchar.INSTANCE)) {
+			return getJsonFromVarchar(object, actualType);
+		}
+		return throwConstraintViolationException(actualType, this);
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType) {
+		return equalsAny(targetType, this, PVarchar.INSTANCE);
+
+	}
+
+	@Override
+	public boolean isCoercibleTo(
+			@SuppressWarnings("rawtypes") PDataType targetType, Object value) {
+		return isCoercibleTo(targetType);
+	}
+
+	@Override
+	public boolean isSizeCompatible(ImmutableBytesWritable ptr, Object value,
+			@SuppressWarnings("rawtypes") PDataType srcType, Integer maxLength,
+			Integer scale, Integer desiredMaxLength, Integer desiredScale) {
+		if (ptr.getLength() != 0 && maxLength != null
+				&& desiredMaxLength != null) {
+			return maxLength <= desiredMaxLength;
+		}
+		return true;
+	}
+
+	@Override
+	public boolean isFixedWidth() {
+		return false;
+	}
+
+	@Override
+	public int estimateByteSize(Object o) {
+		PhoenixJson phoenixJson = (PhoenixJson) o;
+		return phoenixJson.toString().length();
+	}
+
+	@Override
+	public Integer getByteSize() {
+		return null;
+	}
+
+	@Override
+	public int compareTo(Object lhs, Object rhs,
+			@SuppressWarnings("rawtypes") PDataType rhsType) {
+		if (PJsonDataType.INSTANCE != rhsType) {
+			throwConstraintViolationException(rhsType, this);
+		}
+		PhoenixJson phoenixJsonLHS = (PhoenixJson) lhs;
+		PhoenixJson phoenixJsonRHS = (PhoenixJson) rhs;
+		return PVarchar.INSTANCE.compareTo(phoenixJsonLHS.toString(),
+				phoenixJsonRHS.toString(), rhsType);
+	}
+
+	@Override
+	public Object toObject(String value) {
+		byte[] jsonData = Bytes.toBytes(value);
+		return getPhoenixJson(jsonData, 0, jsonData.length);
+	}
+
+	@Override
+	public boolean isBytesComparableWith(
+			@SuppressWarnings("rawtypes") PDataType otherType) {
+		return otherType == PJsonDataType.INSTANCE
+				|| otherType == PVarchar.INSTANCE;
+	}
+
+	@Override
+	public String toStringLiteral(Object o, Format formatter) {
+		if (o == null) {
+			return StringUtil.EMPTY_STRING;
+		}
+		PhoenixJson phoenixJson = (PhoenixJson) o;
+		return PVarchar.INSTANCE.toStringLiteral(phoenixJson.toString(),
+				formatter);
+	}
+
+	@Override
+	public void coerceBytes(ImmutableBytesWritable ptr, Object o,
+			@SuppressWarnings("rawtypes") PDataType actualType,
+			Integer actualMaxLength, Integer actualScale,
+			SortOrder actualModifier, Integer desiredMaxLength,
+			Integer desiredScale, SortOrder expectedModifier) {
```

Comment:
I am ignoring the order of json here since inversion of order will make json invalid.